from gameparts import Board
# Из файла exceptions.py, который лежит в пакете gameparts,
# импортируется класс FieldIndexError.
from gameparts.exceptions import FieldIndexError


# Всё, что в функции, не будет импортироваться,
# но будет выполняться при запуске файла game.py.
def main():
    # Создать игровое поле - объект класса Board.
    game = Board()
    # Отрисовать поле в терминале.
    game.display()

    # Запускается бесконечный цикл.
    while True:
        # В этом блоке содержатся операции, которые могут вызвать исключение.
        try:
            # Тут пользователь вводит координаты ячейки.
            row = int(input('Введите номер строки: '))
            # Если введённое значение меньше нуля или больше или равно
            # field_size (это значение равно трём, оно хранится в модуле
            # parts.py)...
            if row < 0 or row >= game.field_size:
                # ...выбросить исключение FieldIndexError.
                raise FieldIndexError

            column = int(input('Введите номер столбца: '))
            if column < 0 or column >= game.field_size:
                # ...выбросить исключение FieldIndexError.
                raise FieldIndexError

        # Если возникает исключение FieldIndexError...
        except FieldIndexError:
            # ...выводятся сообщения...
            print(
                'Значение должно быть неотрицательным и меньше '
                f'{game.field_size}.'
            )
            print('Пожалуйста, введите значения для строки и столбца заново.')
            # ...и цикл начинает свою работу сначала,
            # предоставляя пользователю ещё одну попытку ввести данные.
            continue

        except ValueError:
            # ...выводятся сообщения...
            print('Буквы вводить нельзя. Только числа.')
            print('Пожалуйста, введите значения для строки и столбца заново.')
            # ...и цикл начинает свою работу сначала,
            # предоставляя пользователю ещё одну попытку ввести данные.
            continue

        # Если возникла ошибка не описанная в блоке try
        except Exception as e:
            print(f'Возникла ошибка: {e}')

        # Если в блоке try исключения не возникло...
        else:
            # ...значит, введённые значения прошли все проверки
            # и могут быть использованы в дальнейшем.
            # Цикл прерывается.
            break

    # Разместить на поле символ по указанным координатам - сделать ход.
    game.make_move(row, column, 'X')
    print('Ход сделан!')
    # Перерисовать поле с учётом сделанного хода.
    game.display()


# Вызов функции с инструкцией
if __name__ == '__main__':
    main()
